src/main/kotlin/org/jetbrains/kotlinx/jupyter/connection.kt

```diff
@@ -150,14 +227,208 @@ class JupyterConnection(val config: KernelConfig) : Closeable {
+        socket.removeCallback(socketCallback)
+    }
+
+    fun updateSessionInfo(message: Message) {
+        val header = message.data.header ?: return
+        header.session?.let { sessionId = it }
+        header.username?.let { username = it }
+        messageId = message.id
+    }
+
+    override fun send(socketName: JupyterSocket, message: RawMessage) {
+        val socket = fromSocketName(socketName)
+        socket.sendRawMessage(message)
+    }
+
+    override fun sendReply(socketName: JupyterSocket, parentMessage: RawMessage, type: String, content: JsonObject, metadata: JsonObject?) {
+        val message = RawMessageImpl(
+            parentMessage.id,
+            JsonObject(
+                mapOf(
+                    "header" to makeJsonHeader(type, parentMessage),
+                    "parent_header" to (parentMessage.header ?: Json.EMPTY),
+                    "metadata" to (metadata ?: Json.EMPTY),
+                    "content" to content
+                )
+            )
+        )
+        send(socketName, message)
+    }
+
+    private val commOpenCallbacks = ConcurrentHashMap<String, CommOpenCallback>()
+    private val commTargetToIds = ConcurrentHashMap<String, CopyOnWriteArrayList<String>>()
+    private val commIdToComm = ConcurrentHashMap<String, CommImpl>()
+    override fun openComm(target: String, data: JsonObject): Comm {
+        val id = UUID.randomUUID().toString()
+        val newComm = processCommOpen(target, id, data)
+
+        // send comm_open
+        iopub.sendMessage(
+            makeSimpleMessage(
+                MessageType.COMM_OPEN,
+                CommOpen(newComm.id, newComm.target)
+            )
+        )
+
+        return newComm
+    }
+
+    fun processCommOpen(target: String, id: String, data: JsonObject): Comm {
+        val commIds = commTargetToIds.getOrPut(target) { CopyOnWriteArrayList() }
+        val newComm = CommImpl(target, id)
+        commIds.add(id)
+        commIdToComm[id] = newComm
+
+        val callback = commOpenCallbacks[target]
+        callback?.invoke(newComm, data)
+
+        return newComm
+    }
+
+    override fun closeComm(id: String, data: JsonObject) {
+        val comm = commIdToComm[id] ?: return
+        comm.close(data, notifyClient = true)
+    }
+
+    fun processCommClose(id: String, data: JsonObject) {
+        val comm = commIdToComm[id] ?: return
+        comm.close(data, notifyClient = false)
+    }
+
+    fun removeComm(id: String) {
+        val comm = commIdToComm[id] ?: return
+        val commIds = commTargetToIds[comm.target]!!
+        commIds.remove(id)
+        commIdToComm.remove(id)
+    }
+
+    override fun getComms(target: String?): Collection<Comm> {
+        return if (target == null) {
+            commIdToComm.values.toList()
+        } else {
+            commTargetToIds[target].orEmpty().mapNotNull { commIdToComm[it] }
+        }
+    }
+
+    fun processCommMessage(id: String, data: JsonObject) {
```

Comment:
There might be some heavy computations, including execution or interpreter thread work, so maybe it's good to add a possibility to indicate such cases in a signature and send `STATUS.BUSY `